relay-pattern/src/wildmatch.rs

```diff
@@ -0,0 +1,605 @@
+use std::num::NonZeroUsize;
+
+use crate::{Literal, Options, Ranges, Token, Tokens};
+
+/// Matches [`Tokens`] against a `haystack` with the provided [`Options`].
+///
+/// This implementation is largely based on the algorithm described by [Kirk J Krauss]
+/// and combining the two loops into a single one and other small modifications to take advantage
+/// of the already pre-processed [`Tokens`] structure and its invariants.
+///
+/// [Kirk J Krauss]: http://developforperformance.com/MatchingWildcards_AnImprovedAlgorithmForBigData.html
+pub fn is_match(tokens: &Tokens, haystack: &str, options: Options) -> bool {
+    match options.case_insensitive {
+        false => is_match_impl::<_, CaseSensitive>(tokens.as_slice(), haystack),
+        true => is_match_impl::<_, CaseInsensitive>(tokens.as_slice(), haystack),
+    }
+}
+
+/// Bundles necessary matchers for [`is_match_impl`].
+trait Matcher {
+    /// Returns the length of the `needle` in the `haystack` if the `needle` is a prefix of `haystack`.
+    fn is_prefix(haystack: &str, needle: &Literal) -> Option<usize>;
```

Comment:
nit: `is_match` has `needle` before `haystack`, I would keep that order everywhere.

Reply:
I'll change the order in `is_match`, this API mirrors the behaviour of `memchr` and works better for the range APIs where you have a single haystack but multiple options for the 'needle'.